src/Bicep.Local.Extension/Rpc/BicepExtensionImpl.cs

```diff
@@ -20,75 +21,61 @@ public BicepExtensionImpl(ILogger<BicepExtensionImpl> logger, ResourceDispatcher
-        => WrapExceptions(async () => Convert(await dispatcher.GetHandler(request.Resource.Type).Save(Convert(request), context.CancellationToken)));
+    public override Task<ResourceResponseBody> CreateOrUpdate(ResourceRequestBody request, ServerCallContext context)
+        => WrapExceptions(async () => Convert(await dispatcher.GetHandler(request.Type).CreateOrUpdate(Convert(request), context.CancellationToken)));
 
-    public override Task<ExtensibilityOperationResponse> PreviewSave(ExtensibilityOperationRequest request, ServerCallContext context)
-        => WrapExceptions(async () => Convert(await dispatcher.GetHandler(request.Resource.Type).PreviewSave(Convert(request), context.CancellationToken)));
+    public override Task<ResourceResponseBody> Preview(ResourceRequestBody request, ServerCallContext context)
+        => WrapExceptions(async () => Convert(await dispatcher.GetHandler(request.Type).Preview(Convert(request), context.CancellationToken)));
 
-    public override Task<ExtensibilityOperationResponse> Get(ExtensibilityOperationRequest request, ServerCallContext context)
-        => WrapExceptions(async () => Convert(await dispatcher.GetHandler(request.Resource.Type).Get(Convert(request), context.CancellationToken)));
+    public override Task<ResourceResponseBody> Get(ResourceReferenceRequestBody request, ServerCallContext context)
+        => WrapExceptions(async () => Convert(await dispatcher.GetHandler(request.Type).Get(Convert(request), context.CancellationToken)));
 
-    public override Task<ExtensibilityOperationResponse> Delete(ExtensibilityOperationRequest request, ServerCallContext context)
-        => WrapExceptions(async () => Convert(await dispatcher.GetHandler(request.Resource.Type).Delete(Convert(request), context.CancellationToken)));
+    public override Task<ResourceResponseBody> Delete(ResourceReferenceRequestBody request, ServerCallContext context)
+        => WrapExceptions(async () => Convert(await dispatcher.GetHandler(request.Type).Delete(Convert(request), context.CancellationToken)));
 
     public override Task<Empty> Ping(Empty request, ServerCallContext context)
         => Task.FromResult(new Empty());
 
-    private static Protocol.ExtensibilityOperationRequest Convert(ExtensibilityOperationRequest request)
-    {
-        return new(
-            new(request.Import.Provider, request.Import.Version, request.Import.Config is { } ? JsonNode.Parse(request.Import.Config) as JsonObject : null),
-            new(request.Resource.Type, request.Resource.Properties is { } ? JsonNode.Parse(request.Resource.Properties) as JsonObject : null));
-    }
+    private Protocol.ResourceRequestBody Convert(ResourceRequestBody request)
+        => new(!string.IsNullOrEmpty(request.Config) ? JsonNode.Parse(request.Config) as JsonObject : null,
+                request.Type,
+                JsonNode.Parse(request.Properties)!.AsObject(), // TODO: Is there a better way to parse to ensure result might not be null?
```

Comment:
It's a risk when parsing untrusted data, so it makes sense that we're forced to handle `null`. My recommendation would be to just throw an exception if it happens.

I also think there's a lot going on in this function, so might be worth simplifying by using a traditional function rather than the expression body syntax:

```csharp
private Protocol.ResourceRequestBody Convert(ResourceRequestBody request)
{
    JsonObject? config = null;
    if (request.Config is {})
    {
        config = JsonNode.Parse(request.Config)?.AsObject()
            ?? throw new ...;
    }

    var properties = JsonNode.Parse(request.Properties)?.AsObject()
        ?? throw new ...;

    return new(config, request.Type, properties, request.ApiVersion);
}
```

Reply:
This was intended as a response to your comment `// TODO: Is there a better way to parse to ensure result might not be null?` FYI

Reply:
yup, I updated the code and will push in the upcoming iteration